enclave/core/sgx/enter.S

```diff
@@ -39,20 +39,50 @@
 oe_enter:
 .cfi_startproc
 
+.get_td:
+
+    // Upon first entry to the enclave, td->base.self in the td_t structure
+    // is not yet initialed. However, the loader in host/sgx/create.c places
+    // the td_t structure as a specific offset from TCS, which we can obtain
+    // via td_from_tcs
+
+    // Save register state to stack
+    push %rax
+    push %rbx
+    push %rcx
+    push %rdi
+    push %rsi
+    push %rdx
+
+    // Get the location of the td_t structure for this thread. This value is
+    // expected to be present in %r11 for the remainder of oe_enter.
+    mov %rbx, %rdi
+    call td_from_tcs
+    mov %rax, %r11
+
+    // Restore register state
+    pop %rdx
```

Comment:
Keep in mind that upon entry, rsp still points to host memory. Therefore the pushes and pops actually operate on host memory which is unsafe.

Reply:
Ah good catch. I'll change it